stat/distuv/binomial.go

```diff
@@ -0,0 +1,184 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package distuv
+
+import (
+	"math"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/mathext"
+	"gonum.org/v1/gonum/stat/combin"
+)
+
+// Binomial implements the binomial distribution, a discrete probability distribution
+// that expresses the probability of a given number of successful Bernoulli trials
+// out of a total of n, each with successs probability p.
+// The binomial distribution has density function:
+//  f(k) = (n choose k) p^k (1-p)^(n-k)
+// For more information, see https://en.wikipedia.org/wiki/Binomial_distribution.
+type Binomial struct {
+	// N is the total number of Bernoulli trials. N must be greater than 0.
+	N float64
+	// P is the probablity of success in any given trial. P must be on [0, 1].
+	P float64
+
+	Src rand.Source
+}
+
+// CDF computes the value of the cumulative distribution function at x.
+func (b Binomial) CDF(x float64) float64 {
+	if x < 0 {
+		return 0
+	}
+	if x >= b.N {
+		return 1
+	}
+	x = math.Floor(x)
+	return mathext.RegIncBeta(b.N-x, x+1, 1-b.P)
+}
+
+// ExKurtosis returns the excess kurtosis of the distribution.
+func (b Binomial) ExKurtosis() float64 {
+	v := b.P * (1 - b.P)
+	return (1 - 6*v) / (b.N * v)
+}
+
+// LogProb computes the natural logarithm of the value of the probability
+// density function at x.
+func (b Binomial) LogProb(x float64) float64 {
+	if x < 0 || x > b.N || math.Floor(x) != x {
+		return math.Inf(-1)
+	}
+	lb := combin.LogGeneralizedBinomial(b.N, x)
+	return lb + x*math.Log(b.P) + (b.N-x)*math.Log(1-b.P)
+}
+
+// Mean returns the mean of the probability distribution.
+func (b Binomial) Mean() float64 {
+	return b.N * b.P
+}
+
+// NumParameters returns the number of parameters in the distribution.
+func (Binomial) NumParameters() int {
+	return 2
+}
+
+// Prob computes the value of the probability density function at x.
+func (b Binomial) Prob(x float64) float64 {
+	return math.Exp(b.LogProb(x))
+}
+
+// Rand returns a random sample drawn from the distribution.
+func (b Binomial) Rand() float64 {
+	// NUMERICAL RECIPES IN C: THE ART OF SCIENTIFIC COMPUTING (ISBN 0-521-43108-5)
+	// p. 294
+	// <http://www.aip.de/groups/soe/local/numres/bookcpdf/c7-3.pdf>
+
+	unifrnd := rand.Float64
+	exprnd := rand.ExpFloat64
+	var rng *rand.Rand
+	if b.Src != nil {
+		rng = rand.New(b.Src)
+		unifrnd = rng.Float64
+		exprnd = rng.ExpFloat64
+	}
+
+	p := b.P
+	if p > 0.5 {
+		p = 1.0 - p
+	}
+	am := b.N * p
+
+	if b.N < 25.0 {
+		// Use direct method.
+		bnl := 0.0
+		for i := 0; i < int(b.N); i++ {
+			if unifrnd() < p {
+				bnl++
+			}
+		}
+		if p != b.P {
+			return b.N - bnl
+		}
+		return bnl
+	}
+
+	if am < 1.0 {
+		// Use rejection method with Poisson proposal.
```

Comment:
Where does this algorithm come from? The link you provided only has the innermost part of this for loop for the `am < 1` case (using the direct poisson).

Reply:
I can add a note in the code, but the Poisson is an asymptotic approximation and not quite good enough for `n` near the cutoff of 25. The experimental variance and ex-kurtosis had greather than `tol=1e-2` absolute error from theoretical predictions. Rejecting a small number of these Possion variates corrects this. The algorithm is simple rejection sampling with some simplifications using Stirling's exansion and the Padé approximant for `log(1+x)` to speed up the computation.

Reply:
The idea behind the correction definitely makes sense, and I was surprised not to see it in the book (similar ideas come up for the Gamma distribution). I am curious if there is a reference, as it would make the code easier to review.

Reply:
I'll check around the interwebs and see if I can find anything.

Reply:
Have you been able to find anything for this?

Reply:
I haven't been able to find previous use of this particular target/proposal pair. I did find a [reference](http://www.stat.yale.edu/~pollard/Courses/241.fall97/Poisson.pdf) that the total deviation in probability is `O(p)` (with a constant factor of 4) and therefore at worst `O(1/n)` on the `am < 1` branch. This sublinear convergence suggests that rejection sampling, which is always guaranteed to be exact and has at worst a < 3% rejection rate, is probably a better choice than simply increasing the cutoff.